picasso/src/main/java/com/squareup/picasso/MarkableInputStream.java

```diff
@@ -127,6 +131,9 @@ private void skip(long current, long target) throws IOException {
   }
 
   @Override public int read(byte[] buffer) throws IOException {
+    if (offset + buffer.length >= limit) {
+      setLimit(2 * (offset + buffer.length));
```

Comment:
I see the problem. I think the solution would be to let the mark(s) to expire once there has been a reset back to the earliest mark.